### ISSUE TITLE
Can programs for a given platform run operations in parallel? 

### ISSUE THREAD

**Issue by mainrs** [roc-lang/roc#6271]
I have written a platform with support for HTTP requests for website crawling. 

Currently a user calls a function called `crawl` with a given start url. And the function returns a response object one can operate on. 

A user can submit a new crawl by calling the crawl function again with a function callback that gets called once the response is available.

But everything is sequential on the roc side of things. Is it possible to make it concurrent? 

For example use generators to submit crawled data or new requests instead of calling functions. And the platform can pull the generator for new instances.

**Comment by Anton-4** [comment]
We have `Task.map2` in the plans as a concurrency primitive. It would allow you to run two tasks in parallel. It will take some time until it's ready though.

As a temporary hack, you could submit a list of urls to your platform and return nothing (`Task {} YourErrorType`), and then use [`Task.loop`](https://github.com/roc-lang/basic-cli/blob/d11fda1c16683e09fac036a1a062aa5889ae8fab/examples/echo.roc#L10) with a [sleep](https://github.com/roc-lang/basic-cli/blob/d11fda1c16683e09fac036a1a062aa5889ae8fab/examples/time.roc) time of your choosing. In that loop you can ask the platform if it finished crawling more urls and process the results as they come in. That processing on the roc side would be single threaded but now the platform could execute requests in parallel.

**Comment by mainrs** [comment]
> We have `Task.map2` in the plans as a concurrency primitive. It would allow you to run two tasks in parallel. It will take some time until it's ready though.

Is it possible to generalize over `n` arguments without having to create a new method each time? Similar to Rust's const generics feature?

**Comment by Anton-4** [comment]
I think so but I asked [on zulip](https://roc.zulipchat.com/#narrow/stream/395097-compiler-development/topic/map2.20to.20mapN/near/421848544) to make sure.

**Comment by Anton-4** [comment]
Yes, Folkert confirmed `mapN` should work by recursively using `map2` to map over N tasks.

**Comment by mainrs** [comment]
> Yes, Folkert confirmed `mapN` should work by recursively using `map2` to map over N tasks.

That would make calling semantics easier for sure. This would still mean that only 2 tasks get executed in parallel, is that correct?

**Comment by Anton-4** [comment]
@folkertdev

**Comment by skyqrose** [comment]
This question is pretty old, so I'm assuming it's no longer relevant and closing the issue. Sorry if your question wasn't answered. If you still have questions about this, please feel free to re-open this issue, open a new issue, or ask on Zulip.
